controller/openebs/cstor.go

```diff
@@ -111,6 +150,532 @@ func (p *Planner) setCStorDefaultsIfNotSet() error {
+			"name": "iscsiadm-lib-crypto",
+			"hostPath": map[string]interface{}{
+				"type": "File",
+				"path": "/lib64/libcrypto.so.1.0.0",
+			},
+		},
+	}
+	libOpeniscsiusrVolume := unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"name": "iscsiadm-lib-openiscsiusr",
+			"hostPath": map[string]interface{}{
+				"type": "File",
+				"path": "/usr/lib64/libopeniscsiusr.so.0.2.0",
+			},
+		},
+	}
+
+	// Append the volume into existing list of volumes
+	volumes = append(volumes, libCryptoVolume.Object, libOpeniscsiusrVolume.Object)
+
+	// Update the volumes in the daemonset object.
+	err = unstructured.SetNestedSlice(daemonset.Object, volumes, "spec", "template", "spec", "volumes")
+	if err != nil {
+		return err
+	}
+
+	// Fetch the containers.
+	containers, err := unstruct.GetNestedSliceOrError(daemonset, "spec", "template", "spec", "containers")
+	if err != nil {
+		return err
+	}
+
+	// updateContainer function will be called for each cotainer to update.
+	// This function will update only "openebs-csi-plugin" container
+	updateContainer := func(obj *unstructured.Unstructured) error {
+		containerName, _, err := unstructured.NestedString(obj.Object, "spec", "name")
+		if err != nil {
+			return err
+		}
+
+		if containerName == ContainerOpenEBSCSIPluginName {
+			// Fetch the volume mounts of the "openebs-csi-plugin" container.
+			volumeMounts, _, err := unstruct.GetSlice(obj, "spec", "volumeMounts")
+			if err != nil {
+				return err
+			}
+
+			// Create the new volume mounts for the "openebs-csi-plugin" container.
+			libCryptoVolumeMount := unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"name":      "iscsiadm-lib-crypto",
+					"mountPath": "/lib/x86_64-linux-gnu/libcrypto.so.1.0.0",
+				},
+			}
+			libOpeniscsiusrVolumeMount := unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"name":      "iscsiadm-lib-openiscsiusr",
+					"mountPath": "/lib/x86_64-linux-gnu/libopeniscsiusr.so.0.2.0",
+				},
+			}
+
+			// Append the new volume mount in the existing volume mounts.
+			volumeMounts = append(volumeMounts, libCryptoVolumeMount.Object, libOpeniscsiusrVolumeMount.Object)
+
+			// Update the volume mounts in the daemonset object.
+			err = unstructured.SetNestedSlice(obj.Object, volumeMounts, "spec", "volumeMounts")
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	// Update the containers.
+	err = unstruct.SliceIterator(containers).ForEachUpdate(updateContainer)
+	if err != nil {
+		return err
+	}
+
+	// Set back the value of the containers.
+	err = unstructured.SetNestedSlice(daemonset.Object,
+		containers, "spec", "template", "spec", "containers")
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// addExtraVolumeMountsForSUSE15 adds extra volume mounts required in openebs-csi-plugin container for the csi
+// and iscsi to work in suse 15 OS.
+func (p *Planner) addExtraVolumeMountsForSUSE15(daemonset *unstructured.Unstructured) error {
```

Comment:
Can we try to avoid these specific functions and try to have one function which will return OS-specific formed volumes and volume structures and one function will fill it in the given daemonset.